File: algorithms/medium/nearest_exit_from_entrance_in_maze.cpp
class Solution {
public:
    int nearestExit(vector<vector<char>>& maze, vector<int>& entrance) {
      queue<array<int, 3>> q;
      q.push({entrance[0], entrance[1], 0});
      int dirs[4][2] = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
      while (!q.empty()) {
        auto [x, y, ct] = q.front(); q.pop();
        if ((x == 0 || x == maze.size() - 1 || y == 0 ||
            y == maze[0].size() - 1) && (x != entrance[0] || y != entrance[1])) {
          return ct;
        }
        for (auto& [dx, dy] : dirs) {
          int nx = x + dx, ny = y + dy;
          if (nx >= 0 && nx < maze.size() && ny >= 0 && ny < maze[0].size() && maze[nx][ny] == '.') {
            q.push({nx, ny, ct + 1});
            maze[nx][ny] = '+';
          }
        }
      }
      return -1;
    }
};

File: algorithms/medium/nearest_exit_from_entrance_in_maze.ex
defmodule Solution do
  @spec nearest_exit(maze :: [[char]], entrance :: [integer]) :: integer
  def nearest_exit(maze, [x, y]) do
    m = maze |> length()
    n = maze |> hd() |> length()

    g =
      maze
      |> Enum.with_index()
      |> Enum.reduce(%{}, fn {row, i}, acc ->
        row
        |> Enum.with_index()
        |> Enum.reduce(acc, fn {x, j}, acc ->
          Map.put(acc, {i, j}, x)
        end)
      end)

    bfs(:queue.from_list([{x, y, 0}]), {g, m, n}, MapSet.new(), [x, y])
  end

  def bfs(q, {g, m, n} = grid, vistied, entrance) do
    if :queue.is_empty(q) do
      -1
    else
      {x, y, ct} = :queue.head(q)
      q = :queue.tail(q)

      cond do
        exit?(grid, x, y) && [x, y] != entrance ->
          ct

        {x, y} in vistied ->
          bfs(q, grid, vistied, entrance)

        true ->
          vistied = MapSet.put(vistied, {x, y})

          [{-1, 0}, {0, 1}, {1, 0}, {0, -1}]
          |> Enum.reduce(q, fn {dx, dy}, acc ->
            if valid?(grid, x + dx, y + dy) do
              :queue.snoc(acc, {x + dx, y + dy, ct + 1})
            else
              acc
            end
          end)
          |> bfs(grid, vistied, entrance)
      end
    end
  end

  def exit?({_, m, n}, x, y) do
    x == 0 || x == m - 1 || y == 0 || y == n - 1
  end

  def valid?({g, m, n}, x, y) do
    x >= 0 && x < m && y >= 0 && y < n && g[{x, y}] == ?.
  end
end
